Introduce stage testing

ondrejbudai opened the pull request:
Greetings, this a pretty huge PR, grab yourself a cup of coffee and hear my words!

This PR introduces stage tests: The stage testing is based on an output from the tree-diff tool. During one test two pipelines are run and their outputs are compared using tree-diff. The diff is then compared with expected diff included in the repository.

There are currently some issues with this PR I would like to discuss. **One of them is blocker and we shall not merge this PR before it's solved.**

## Blocker
### SELinux
SELinux works differently on Fedora host than on Ubuntu host. On Fedora users and timezone stages change selinux labels on some files. On Ubuntu it doesn't happen. See also #89

We need to investigate what's exactly happening: Why Fedora changes the labels and why it doesn't happen on Ubuntu.

**Currently, there's a diff from Fedora, which causes the tests to fail on CI (which is running Ubuntu)**. But if you run the tests on Fedora, everything is alright. I will investigate images built on Ubuntu to see how selinux labels are applied there.

## Discussions
### How should diffs be "deep"
I've changed the behaviour of `tree-diff` to dump the whole added/deleted subtree instead of just its root. This is imho good, because we can detect more stuff. However, we are not currently not dumping anything else than file names. Therefore we are not able to check for content/permissions/selinux when file is added (or deleted, but doesn't make any sense, I guess?).

Should we dump more information about added (deleted) files?

## Testing of dnf stage
The dnf stage test is weird, the diff is huge... What can we do about that? Install only a small package on top of @<!-- -->Core image (pipeline with dnf stage installing @<!-- -->Core packages)? Currently, it feels to me like we are testing the dnf itself...

## Preserving store between tests
Preserving store between tests sounds bad. Unfortunately, it takes **1 hour (!!)** to run the current 9 stage tests without preserving the store. So, I've changed the tests to preserve the store between them. This works right now, but can have unexpected consequences in future mainly due to not ideal reproducibility of osbuild pipelines. Possible solutions:

1) Just don't install @<!-- -->Core every time!
    Yeah, that would probably speed up the tests drastically. But do we want to test the other stages on empty image? Doesn't sound like a realistic usecase of stage...

2) Introduce some kind of base image, which all stage tests should use (core packages + selinux maybe?).
   Seems like an interesting idea for me, the only downside is that test framework will be probably a little bit more complex? Or not?

3) Leave as it is in current PR
   Yeah, it works, but I'm scared it will break sometime and debugging might be hard.

4) 1 hour test? We don't mind!
   I mind! It's just too much.

teg commented:
Great stuff!

As written in the chat, note that Ubuntu kernels don't have SELinux support. We may still be able to label our files using our userspace tool, but if running tools causes relabelling (which sounds wrong), then I would expect that to not happen on Ubuntu.

For added/deleted trees, I totally agree that we should export all the info about added files, but for the deleted ones I guess only the root of the subtree is sufficient, no?

Agreed, don't test @Core, test just a minimal set.

I think also that we should test stages on minimal/empty images. Some may not work at the moment, but I think (as mentioned before) that stages should be "write only" to the extent possible, so in general should not require anything particular to be installed. Mabye we need a very minimal dnf stage that puts in place the filesystem etc, but that should be it.

General comment on SELinux: we may in practice want to always do the SELinux stage last, so not care so much if applying stages afterwords changes labels, but obviously that is not ideal.

Of your four proposals, I would suggest number 2, but with maybe even without SELinux if we change the stage to read the labels from the build root, not the target (in which case the selinux package of course must be installed there instead).

teg commented on tree-diff:
```diff
@@ -1,10 +1,22 @@
+        hasher.update(buf)
+        buf = os.read(fd, BLOCK_SIZE)
+
+    return f"sha256-{hasher.hexdigest()}"
```
`sha512:` is the usual prefix, rather than `sha512-`.

ondrejbudai replied:
Changed.

teg commented on tree-diff:
```diff
@@ -1,10 +1,22 @@
 
 
+def hash_file(fd):
+    BLOCK_SIZE = 65536
```
Did you measure this? I used 4k elsewhere (one page). If it makes a difference, we should use the same everywhere :)

ondrejbudai replied:
I did no measures, I've changed it to 4096 to be consistent. We can do some benchmarks later if needed. :)

teg commented on test/osbuildtest.py:
```diff
@@ -34,4 +34,4 @@ def run_osbuild(self, pipeline):
 
         result = json.loads(r.stdout)
-        return result["tree_id"], result["output_id"]
+        return f"{self.store}/refs/{result['tree_id']}", f"{self.store}/refs/{result['output_id']}"
```
Hm, I'm not 100% sure about this. I quite liked the current outputs, as they were about the semantics of what we produced, and could potentially be used for other things than just dereferencing the results.

larskarlitski replied:
Agreed. This is only test code, but I liked the symmetry of it returning something similar to what `osbuild --json` on the command line.

Is constructing the part inconvenient or impossible somewhere?

ondrejbudai replied:
I liked the fact that subclasses of osbuildtest.TestCase don't need to know anything about `self.store`. But yeah, it makes sense to return the ids, I will make a change.

teg commented on tree-diff:
```diff
@@ -39,7 +39,7 @@ def selinux_diff(path1, path2, path, differences):
     if label1 != label2:
         props = differences.setdefault(path, {})
-        props["selinux"] = [os.fsdecode(label1), os.fsdecode(label2)]
+        props["selinux"] = [label1.strip('\n\0'), label2.strip('\n\0')]
```
Hm, maybe we should just treat these as byte strings?

ondrejbudai replied:
What do you mean by byte strings?

teg replied:
Ugh, I misread the code. Ignore me :)

larskarlitski commented on .travis.yml:
```diff
@@ -31,7 +31,10 @@ jobs:
+      script: sudo env "PATH=$PATH" python3 -m test --case locale --build-pipeline samples/build-from-yum.json
+    - name: stage-tests
+      before_install: sudo apt-get install -y systemd-container yum
+      script: sudo env "PATH=$PATH" "OSBUILD_TEST_BUILD_PIPELINE=samples/base-from-yum.json" python3 -m unittest test.test_stages
```
Please use `samples/build-from-yum.json` instead. It's quicker. I've removed the other one, because it's confusing :)

ondrejbudai replied:
What do you mean by quicker? When I'm using `base-from-yum` the pipeline works like this:

1) build f27
2) build f30
3) build f30

On the contrary with `build-from-yum`:

1) build f27
2) build f30

I think the first pipeline should be used, it sounds more reasonable to build the final tree by the same Fedora version. Currently, this is solved by using build step (F30) in each of test pipelines AND using build-pipelined argument (F27). IMO, this is confusing (which build pipeline goes first?) and it means I need to put the buildstep in each test pipeline, which is cluttering them a lot.

What do you think?

larskarlitski replied:
I meant it's quicker because it has one stage instead of two. 

The pipeline you're building *always* needs to include its build pipeline, because its hash must include all the tools that were used to assemble the final image. The `--build-pipeline` parameter is there to get a host system into something that we know can produce the build pipeline. It is always run first. The Sorry if this is confusing.

The first case is indeed the one that we should use (for the reason you mention). The external pipeline makes a f27, to make it possible to build the "real" pipeline (f30 built on f30) on Ubuntu.

Does that make sense?

ondrejbudai replied:
I thought that `--build-pipeline` gets also included into the tree id. But I was wrong, thanks for the explanation. :) Therefore I added the build stage to all the test pipelines. :)

larskarlitski commented on test/osbuildtest.py:
```diff
@@ -15,21 +15,102 @@ class TestCase(unittest.TestCase):
-        shutil.rmtree(self.store)
+    @classmethod
+    def tearDownClass(cls):
+        shutil.rmtree(cls.store)
```
I'm doing the same for testing image-info, but I think we need to be a bit careful here, where we actually test osbuild itself. Tests should be as independent from each other as possible.

I'm not entirely against it. If we go this way, please include a comment pointing this out as the doc string for this test class.

larskarlitski commented on test/osbuildtest.py:
```diff
@@ -15,21 +15,102 @@ class TestCase(unittest.TestCase):
             r = subprocess.run(osbuild_cmd, encoding="utf-8", stdout=subprocess.PIPE, check=True)
+        except subprocess.CalledProcessError as e:
+            print(e.stdout)
+            raise e from None
```
Neat idea. This will help spot problems on travis.

larskarlitski commented on test/osbuildtest.py:
```diff
@@ -15,21 +15,102 @@ class TestCase(unittest.TestCase):
+            diff = ('\n' + '\n'.join(difflib.ndiff(
+                pprint.pformat(tree_diff1).splitlines(),
+                pprint.pformat(tree_diff2).splitlines())))
+            raise AssertionError(diff) from None
```
`raise from None` removes the error message you've put in the `AssertionErrors` above. Is this intentional?

ondrejbudai replied:
Fixed by introducing `TreeAssertionError` and removing the ugly try-except-reraise thing.

larskarlitski commented on test/test_stages.py:
```diff
@@ -0,0 +1,36 @@
+def init_tests():
+    test_dir = 'test/stages_tests'
+    for test in os.listdir(test_dir):
+        setattr(TestDescriptions, f"test_{test}", generate_test_case(f"{test_dir}/{test}"))
```
I prefer using `TestCase.subTest()` for parameterized tests. No need to change to that, just wanted to mention it.

ondrejbudai replied:
`setUp` and `tearDown` don't work with `subTest()`

larskarlitski replied:
Ah! Seriously?! Thanks for pointing that out.

ondrejbudai replied:
I've tested this once again today's morning and yeah, it works this way. But I guess it's a correct behaviour, if setUp got called before each subTest, it would got called twice - once before test method and then before subtest, which would be imho really confusing.

larskarlitski commented on tree-diff:
```diff
@@ -77,6 +85,13 @@ def diff_aux(dir_fd1, dir_fd2, path, report):
+                    try:
+                        child_fd = os.open(dirent.name, os.O_DIRECTORY, dir_fd=dir_fd1)
+                    except OSError:
+                        continue
```
Which `OSError` do you want to ignore? I feel like this except handler might be too broad. Maybe something like [this](https://github.com/osbuild/osbuild/blob/master/osbuild/objectstore.py#L16)?

larskarlitski commented:
> The dnf stage test is weird, the diff is huge... What can we do about that? Install only a small package on top of @Core image (pipeline with dnf stage installing @Core packages)? Currently, it feels to me like we are testing the dnf itself...

I don't think we need to include it. There's other ways to test it without bloating the size of the repo.

> Preserving store between tests sounds bad

Ah sorry I didn't read this before putting my comment. I vote (3). It's the easiest solution. We are not testing that the osbuild *store* works, but that stages are doing the right thing. We have other ways to test the store.

ondrejbudai commented:
> I think also that we should test stages on minimal/empty images. Some may not work at the moment, but I think (as mentioned before) that stages should be "write only" to the extent possible, so in general should not require anything particular to be installed. Mabye we need a very minimal dnf stage that puts in place the filesystem etc, but that should be it.

Idea with "write only" stages is certainly great. But in real use case we will probably always write e.g. users or firewall settings after dnf run. That's why I want to use full @<!-- -->core as base in all tests.

ondrejbudai commented:
@teg @larskarlitski 

This version is imho ready to be merged. There's a brief description how we solved my discussion points:

- selinux - we consider the selinux labels as garbage during the pipeline run. There should always be selinux stage at the end of pipeline to do a global relabel.
- deepness of diffs - we will add metadata to `added_files`. In `deleted_files` it makes sense to include only the root of deleted subtree into the diff. This PR is good enough even without this behaviour (just the tests are as not thorough as they can be). I will create an issue for this change.
- dnf stage - I've removed it for the time being. It tested more dnf behaviour than our code. We should think of better ways how to test the dnf stage.
- preserving the store during tests - I've changed the behaviour to preserve the store through all tests in one test class. This may make some problems in the future, but currently works just fine, really speeds up the time times and it's the most simple and KISS solution.

teg commented:
@ondrejbudai this looks really great! I think there is still one outstanding comment about OSError exception from Lars. Otherwise, I'm happy to merge this.

larskarlitski commented:
> @ondrejbudai this looks really great! I think there is still one outstanding comment about OSError exception from Lars. Otherwise, I'm happy to merge this.

Can be done as a follow up.

Thanks Ondřej! This is really nice :tada:
